io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -1079,130 +1090,290 @@ traceMany ((time, tid, event):ts) trace =
+          go ctl' read written' writtenSeq' nextVid (k x)
+
+        OrElseRightFrame k writtenOuter writtenOuterSeq ctl' -> do
+          -- Commit the TVars written in this sub-transaction that are also
+          -- in the written set of the outer transaction
+          traverse_ (\(SomeTVar tvar) -> commitTVar tvar)
+                    (Map.intersection written writtenOuter)
+          -- Merge the written set of the inner with the outer
+          let written'    = Map.union written writtenOuter
+              writtenSeq' = filter (\(SomeTVar tvar) ->
+                                      tvarId tvar `Map.notMember` writtenOuter)
+                                    writtenSeq
+                         ++ writtenOuterSeq
+          -- Continue with the k continuation
+          go ctl' read written' writtenSeq' nextVid (k x)
+
+      ThrowStm e -> do
+        -- Revert all the TVar writes
+        traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+        return (StmTxAborted (toException e))
+
+      Retry -> case ctl of
+        AtomicallyFrame -> do
+          -- Revert all the TVar writes
+          traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+          -- Return vars read, so the thread can block on them
+          return (StmTxBlocked (Map.elems read))
+
+        OrElseLeftFrame b k writtenOuter writtenOuterSeq ctl' -> do
+          -- Revert all the TVar writes within this orElse
+          traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+          -- Execute the orElse right hand with an empty written set
+          let ctl'' = OrElseRightFrame k writtenOuter writtenOuterSeq ctl'
+          go ctl'' read Map.empty [] nextVid b
+
+        OrElseRightFrame _k writtenOuter writtenOuterSeq ctl' -> do
+          -- Revert all the TVar writes within this orElse branch
+          traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+          -- Skip the continuation and propagate the retry into the outer frame
+          -- using the written set for the outer frame
+          go ctl' read writtenOuter writtenOuterSeq nextVid Retry
+
+      OrElse a b k -> do
+        -- Execute the left side in a new frame with an empty written set
+        let ctl' = OrElseLeftFrame b k written writtenSeq ctl
+        go ctl' read Map.empty [] nextVid a
+
+      NewTVar x k -> do
+        v <- execNewTVar nextVid x
+        go ctl read written writtenSeq (succ nextVid) (k v)
+
+      ReadTVar v k
+        | tvarId v `Map.member` read -> do
+            x <- execReadTVar v
+            go ctl read written writtenSeq nextVid (k x)
+        | otherwise -> do
+            x <- execReadTVar v
+            let read' = Map.insert (tvarId v) (SomeTVar v) read
+            go ctl read' written writtenSeq nextVid (k x)
+
+      WriteTVar v x k
+        | tvarId v `Map.member` written -> do
+            execWriteTVar v x
+            go ctl read written writtenSeq nextVid k
+        | otherwise -> do
+            saveTVar v
```

Comment:
We only use `saveTVar` once here (I assume that `written` can only grow during the execution).  This suggest we don't need a linear undo stack, I think `OR3` from the `stm` paper is compatible with this claim.

Reply:
@dcoutts looking more closely I convienced myself that we need to use a linear stack. When we push the `OrElseRightFrame` we continue with an empty set of `written` variables. So next write will push a variable onto the undo stack.